syntax = "proto3";

message registerUser {
    string username = 1;
    string password = 2;
    requestReply reply = 3;
}

message loginUser {
    string username = 1;
    string password = 2;
    requestReply reply = 3;
}

message deleteUser {
    string username = 1;
    string from_user = 2;
    string password = 3;
    requestReply reply = 4;
}

message dumpUsers {
    requestReply reply = 1;
}

message filterUsers {
    string wildcard = 1;
    requestReply reply = 2;
}

message messageUser {
    string username = 1;
    string from_user = 2;
    string m = 3;
    requestReply reply = 4;
}

message requestReply {
    string reply = 1; //repeated string usersList
}

service UserTable {
    rpc RegisterUser (registerUser) returns (requestReply);
    rpc LoginUser (loginUser) returns (requestReply);
    rpc DeleteUser (deleteUser) returns (requestReply);
    
    rpc DumpUsers (dumpUsers) returns (requestReply);
    rpc FilterUsers (filterUsers) returns (requestReply);
    rpc MessageUser (messageUser) returns (requestReply);
    rpc SubscribeMessages (requestReply) returns (stream messageUser) {}
}